### PR TITLE
Cleanup of LUTs

### PR DESCRIPTION
**Pull request by aryd**
#### PR description:

The updates in this PR consolidates the functionality of the LUTs in the class TrackletLUT. Code to generate LUTs has now been migrated to this class. In addition to this change a number of changes has been made to remove hard coded numbers for layers/disks and seeds. These numbers have been replaced by enums.

#### PR validation:

These changes have been validated to show that the results don't change and that the same files are written out. 

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

File: L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+
+    double rmin = std::min({r1, r2, r3, r4, r5, r6, r7, r8});
+    double rmax = std::max({r1, r2, r3, r4, r5, r6, r7, r8});
+
+    int NBINS = settings_.NLONGVMBINS() * settings_.NLONGVMBINS() / 2;
+
+    double rmindisk = settings_.rmindiskvm();
+    double rmaxdisk = settings_.rmaxdiskvm();
+
+    if (iseed == 6)
```
Comment (skinnari) [inline L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc]: @aryd could these use the new seed enum too?

File: L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+    // xxx|yy|z|rrr
+    // xxx is the delta r window
+    // yy is the r bin yy is three bits for overlaps
+    // z is flag to look in next bin
+    // rrr fine r bin
+    // NOTE : this encoding is not efficient z is one if xxx+rrr is greater than 8
+    //        and xxx is only 1,2, or 3
+    //        should also reject xxx=0 as this means projection is outside range
+
+    bool overlap = iseed == 6 || iseed == 7 || iseed == 10;
```
Comment (skinnari) [inline L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc]: similar here?

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc [hunk 3]
```diff
@@ -128,17 +169,28 @@ void TrackletProcessor::addInput(MemoryBase* memory, string input) {
...
-    buildLUT();  //need iAllStub_ set before building the table //FIXME should be in initiall
 
+    //cout << getName() << " " << nbitsfinephidiff_<<endl;
+    unsigned int iTP = getName()[7]-'A';
+    
+    
+    pttableinner_.initTPlut(true, iSeed_, layerdisk1_, layerdisk2_, nbitsfinephidiff_, iTP); 
+    pttableouter_.initTPlut(false, iSeed_, layerdisk1_, layerdisk2_, nbitsfinephidiff_, iTP); 
+
+    //need iAllStub_ set before building the table //FIXME should be in initiall
```
Comment (skinnari) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc]: is this intended to be left here?

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc [hunk 4]
```diff
@@ -128,17 +169,28 @@ void TrackletProcessor::addInput(MemoryBase* memory, string input) {
       if (tmp->getName()[11] == 'W')
         iAllStub_ = 3;
     }
-    buildLUT();  //need iAllStub_ set before building the table //FIXME should be in initiall
 
+    //cout << getName() << " " << nbitsfinephidiff_<<endl;
```
Comment (skinnari) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc]: need to remove this line..

File: L1Trigger/TrackFindingTracklet/src/VMRouter.cc [hunk 5]
```diff
@@ -14,15 +14,38 @@ using namespace std;
...
+  unsigned int region = name[9]-'A';
+  assert(region < settings_.nallstubs(layerdisk_));
+  
   overlapbits_ = 7;
   nextrabits_ = overlapbits_ - (settings_.nbitsallstubs(layerdisk_) + settings_.nbitsvmme(layerdisk_));
 
-  vmrtable_.init(layerdisk_, name);
+  meTable_.initVMRTable(layerdisk_, TrackletLUT::VMRTableType::me, region);                    //used for ME and outer TE barrel
+
+  //if (layerdisk_>= N_LAYER) {
```
Comment (skinnari) [inline L1Trigger/TrackFindingTracklet/src/VMRouter.cc]: delete line?

File: L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc (hunk 1 above)
Comment (aryd) [reply L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc]: Yes - have changed, these were missed.

File: L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc (hunk 2 above)
Comment (aryd) [reply L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc]: Also fixed.

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc (hunk 3 above)
Comment (aryd) [reply L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc]: Yes, code can not be moved as written now, some substantial changes would be needed. I modified the comment a bit.

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc (hunk 4 above)
Comment (aryd) [reply L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc]: Fixed.

File: L1Trigger/TrackFindingTracklet/src/VMRouter.cc (hunk 5 above)
Comment (aryd) [reply L1Trigger/TrackFindingTracklet/src/VMRouter.cc]: Yes - removed now.

**Comment by aryd** [comment]
These comments are now addressed.

 -Anders

Anders Ryd
***@***.******@***.***>



On May 6, 2021, at 8:53 AM, Louise Skinnari ***@***.******@***.***>> wrote:



@skinnari commented on this pull request.

________________________________

In L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc<https://github.com/cms-L1TK/cmssw/pull/79#discussion_r627389546>:

> +    double r3 = (r + 0.5 * dr) * (zmean + z0cut) / (z - 0.5 * dz + z0cut);
+    double r4 = (r - 0.5 * dr) * (zmean - z0cut) / (z - 0.5 * dz - z0cut);
+    double r5 = (r + 0.5 * dr) * (zmean - z0cut) / (z + 0.5 * dz - z0cut);
+    double r6 = (r - 0.5 * dr) * (zmean + z0cut) / (z + 0.5 * dz + z0cut);
+    double r7 = (r + 0.5 * dr) * (zmean - z0cut) / (z - 0.5 * dz - z0cut);
+    double r8 = (r - 0.5 * dr) * (zmean + z0cut) / (z - 0.5 * dz + z0cut);
+
+    double rmin = std::min({r1, r2, r3, r4, r5, r6, r7, r8});
+    double rmax = std::max({r1, r2, r3, r4, r5, r6, r7, r8});
+
+    int NBINS = settings_.NLONGVMBINS() * settings_.NLONGVMBINS() / 2;
+
+    double rmindisk = settings_.rmindiskvm();
+    double rmaxdisk = settings_.rmaxdiskvm();
+
+    if (iseed == 6)


@aryd<https://github.com/aryd> could these use the new seed enum too?

________________________________

In L1Trigger/TrackFindingTracklet/src/TrackletLUT.cc<https://github.com/cms-L1TK/cmssw/pull/79#discussion_r627389715>:

> +    if (rbin2 >= NBINS)
+      rbin2 = NBINS - 1;
+    if (rbin1 < 0)
+      rbin1 = 0;
+
+    // This is a 9 bit word:
+    // xxx|yy|z|rrr
+    // xxx is the delta r window
+    // yy is the r bin yy is three bits for overlaps
+    // z is flag to look in next bin
+    // rrr fine r bin
+    // NOTE : this encoding is not efficient z is one if xxx+rrr is greater than 8
+    //        and xxx is only 1,2, or 3
+    //        should also reject xxx=0 as this means projection is outside range
+
+    bool overlap = iseed == 6 || iseed == 7 || iseed == 10;


similar here?

________________________________

In L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc<https://github.com/cms-L1TK/cmssw/pull/79#discussion_r627390254>:

>
+    //cout << getName() << " " << nbitsfinephidiff_<<endl;
+    unsigned int iTP = getName()[7]-'A';
+
+
+    pttableinner_.initTPlut(true, iSeed_, layerdisk1_, layerdisk2_, nbitsfinephidiff_, iTP);
+    pttableouter_.initTPlut(false, iSeed_, layerdisk1_, layerdisk2_, nbitsfinephidiff_, iTP);
+
+    //need iAllStub_ set before building the table //FIXME should be in initiall


is this intended to be left here?

________________________________

In L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc<https://github.com/cms-L1TK/cmssw/pull/79#discussion_r627390334>:

>
+    //cout << getName() << " " << nbitsfinephidiff_<<endl;


need to remove this line..

________________________________

In L1Trigger/TrackFindingTracklet/src/VMRouter.cc<https://github.com/cms-L1TK/cmssw/pull/79#discussion_r627390765>:

>    overlapbits_ = 7;
   nextrabits_ = overlapbits_ - (settings_.nbitsallstubs(layerdisk_) + settings_.nbitsvmme(layerdisk_));

-  vmrtable_.init(layerdisk_, name);
+  meTable_.initVMRTable(layerdisk_, TrackletLUT::VMRTableType::me, region);                    //used for ME and outer TE barrel
+
+  //if (layerdisk_>= N_LAYER) {


delete line?

**Review by skinnari — APPROVED** [review]

**Pull request merged**